### DISCUSSION TITLE
How to use PotatoNV [Unofficial]

### DISCUSSION THREAD

**Discussion by ThomasHeinrichSchmidt** [kitsuned/PotatoNV#24] [Show and tell]
# How to use PotatoNV

## Unlocking the boot loader 
_DISCLAIMER: You will void your device warranty. I have successfully done this guide but I am not responsible for any malfunctions or any damage to the device. Please read carefully every single step and everything will be successful._ 

Unlocking puts your device in a potentially unsafe state in terms of vulnerability to malware.
Incidentally, unlocking will not change the installed software and will not damage your data. 
Back up anything you may need to be safe. 

After unlocking the device will show a message before booting but will otherwise start normally as before. 
![PHONE Booting unlocked](https://user-images.githubusercontent.com/8068047/106356571-f2d30e80-6300-11eb-964f-7f54e7a75381.jpg)

### Install ADB and HUAWEI USB COM drivers
Install HUAWEI USB COM drivers (using "Huawei drivers testpoint.rar" from https://files.dc-unlocker.com/share.html?v=share/18B15B9D02C945A79B1967234CECB423).
This installs without any problems, the drivers are signed by Huawei, confirmed by VirusTotal and Windows.
Open up Device Manager to check that under ports (COM & LPT) connection changes from 'USB SER' to 'USB (COM)'
Uninstall the HUAWEI drivers when finished or else phone won't be recognized properly by PC later!
### Remove the cover from the device: 
Heat the cover on the back of the phone with a hairdryer and carefully peel cover along the small gap using a plastic card or guitar pick 
![PHONE open back cover](https://user-images.githubusercontent.com/8068047/106356851-e354c500-6302-11eb-9fdf-817a4f394fc8.jpg)

### Find where the testpoint is located. 
Sample search phrase "HUAWEI P8 Lite 2017 PRA LA1 testpoint" will result in several hits with with suitable photos to show you the location of the testpoint. In my case it is enough to open the back cover of the phone, you do not need to remove any cables or covers.

### ❋ Init 
Turn off the device and unplug the USB cable

### Short testpoint and iron shield (ground) on the motherboard
Solder a piece of flexible cable to a rigid cable, clamp some wires of the flexible cable to the external part of the UBS connector by connecting it to the phone.
![Two-ended cable](https://user-images.githubusercontent.com/8068047/106356907-2a42ba80-6303-11eb-9638-795b7c362a53.jpg)
You should use a multimeter or resistance meter to check if there is a conductive connection with the outside of the USB connector on the PC side, i.e. if your cable connection to ground is good, 
![Open phone, wire connected to USB ground](https://user-images.githubusercontent.com/8068047/106357009-cff62980-6303-11eb-84ad-43c238c71085.jpg) 
![Clamp flexible cable onto the external part of the UBS connector](https://user-images.githubusercontent.com/8068047/106357023-e4d2bd00-6303-11eb-80f8-3be9f9a25085.jpg)
   Touch the test point with the other stiff end of the wire 
### Connect the Phone to the PC 
using the USB connector, while still touching the test point with the wire 
### Now turn on the phone 
by pushing the power button for 10 seconds, while keeping the cable connected at both ends.
Make sure there’s nothing on the device’s screen. If a charging animation appears go to step **❋ Init**

_I had problems to watch the (front) screen while holding the cable connected on the background of the device, so I did not see what happened (maybe that's the reason why my first attempt failed). But I just tried again from step ❋ Init and it worked the second time._

### Launch PotatoNV 
and select "HUAWEI USB COM 1.0" in the device list (was already automatically chosen)
Select bootloader - see table to help you make a choice. If the device is not in the table, then just try different bootloaders (this is safe).
For my P8 Lite 2017 (PRA-LX1) PotatoNV chose "Kirin 659(A)" by default what was OK.
Fields SN and BOARDID are optional, just leave them blank
Unlock code - this is the future bootloader code, you can change it to any string, the code must contain 16 characters
Do not forget to write down the bootloader unlock code PatatoNV generates (you may need it while trying to install the new system or using eRecovery
### Press the Start button on PotatoNV.
PotatoNV shows:
```
PotatoNV v2.0.1
User manual: https://kutt.it/pnv-en
--> Flashing bootloader
Verifying images...
Verification passed!
Uploading hisi659a bootloader
- xloader
- fastboot
Bootloader uploaded
Waiting for any device...
--> Reading information
Connecting to fastboot device...
- Serial number: 1234567890123456
- Board ID: ABCD567890123456
- Model: PRA-LX1
- Build number: PRA-LX1 8.0.0.410(C432)
- FBLOCK state: unlocked
--> Updating NVME
- Writing FBLOCK state
- Writing User key
Update done!
Rebooting...
```   
![PotatoNV finished (success)](https://user-images.githubusercontent.com/8068047/106357210-f6689480-6304-11eb-97a2-368085476b4c.jpg)

If everything goes well, the phone will reboot and show the following text: "Your device has been unlocked, ..." during boot, see first picture "PHONE Booting unlocked" at top.
### Verify that everything worked out
first you need to enable USB-debugging in the development options of your device then type in command window:
```
      C:\> adb devices
         List of devices attached
         1234567890123456         device
      C:\> fastboot oem get-bootinfo
         < waiting for any device >
         ^C        -- does not work 
      C:\> adb reboot bootloader
                      -- phone shows: PHONE Unlocked (red) / FRP Unlock (red)   [FRP=Factory Reset Protection]
      C:\>fastboot oem get-bootinfo
         ...
         **(bootloader)  unlocked**   
         OKAY [  0.011s]
         finished. total time: 0.013s
      C:\> fastboot reboot
         rebooting...
         finished. total time: 0.026s
``` 
![PHONE unlocked](https://user-images.githubusercontent.com/8068047/106357337-de454500-6305-11eb-83e4-524b3223a5ff.jpg)

### Uninstall HUAWEI USB COM drivers 
To be able to access the phone via USB (to copy files, run ADB etc.) you need to uninstall the HUAWEI USB COM drivers using "DriverUninstall.exe".
Once the uninstall is finished it shows a strange "Restart" message box 
![Huawei drivers testpoint Uninstall finished](https://user-images.githubusercontent.com/8068047/106357385-19e00f00-6306-11eb-8e1c-3a3aa6dce09e.jpg)

If you press OK, your PC will restart (what is probably necessary to actually remove the driver).

**Comment by dnitto** [comment]
Hello good Morning
I have the huawei mate 9 mha l09 I should connect my phone to the pc open the program and start just that or make that connection that you see with the usb cable.

**Reply by ThomasHeinrichSchmidt** [reply]
You need to find and connect the testpoint for your device to ground during the procedure, as described above, otherwise PotatoNV cannot work (to find testpoint see e.g. [Video, from 0:08, watch muted](https://www.youtube.com/watch?v=k_CoUU5nj3U))

**Reply by dnitto** [reply]
Ok thanks

**Reply by affineer** [reply]
hello, it tried. but could NOT succeed. Here is the potatoNv log.

BYU6V9EKEGCFU5GQ



PotatoNV v2.0.1
User manual: https://kutt.it/pnv-en
--> Flashing bootloader
Verifying images...
Verification passed!
Uploading hisi659a bootloader
- xloader
- fastboot
Bootloader uploaded
Waiting for any device...
--> Reading information
Connecting to fastboot device...
- Serial number: WCY4C18603001426
- Board ID: 4C4XNP1862004346
- Model: ANE-LX2
- Build number: ANE-LX2 9.1.0.353(ZAFC185E4R1P8)
- FBLOCK state: unlocked
--> Updating NVME
- Writing FBLOCK state
Failed to set prop:


Where am I wrong?

**Reply by ThomasHeinrichSchmidt** [reply]
I don't know if you are wrong, everything looks good just before writing the User key,
If successful you would have seen - "Writing User key" instead of "Failed to set prop:"
When you look into `PotatoNV-next\Core.cs` you will see 
```
       private void WriteNVME()
        {
            SetNVMEProp("FBLOCK", new[] { (byte)(args.DisableFBLOCK ? 0 : 1) }, "FBLOCK state");
            SetNVMEProp("USRKEY", GetSHA256(args.UnlockCode), "User key");   
            ... 
```
I do not know why the first of these two calls was successful but the second failed. Maybe you lost the connection to the device just at this point?

**Reply by affineer** [reply]
thanks for the reply. 
My mobile is stuck at bootloop. but can enter fastboot, (can't boot to recovery)
I can not use it to toggle OEM UNLOCK to "YES".

maybe issue is with OEM UNLOCK being disabled in Developer Options.

Can I use Fastboot to toggle OEM UNLOCK to "YES"

**Reply by ThomasHeinrichSchmidt** [reply]
I did not switch on `OEM Unlocking` in developer options before using PotatoNV, I actually do not remember if this option existed on my phone anyway. You can use `fastboot oem unlock <PotatoNV-unlock-code>` to unlock the bootloader, if you have the code correct, like this:
```
C:\>fastboot reboot
       rebooting...
       finished. total time: 0.026s
C:\>adb devices
       List of devices attached
       XXXXXXXXXXXXXXX        device
C:\> adb reboot bootloader
C:\> fastboot oem unlock <PotatoNV-unlock-code>
```

**Reply by affineer** [reply]
C:\Program Files (x86)\Minimal ADB and Fastboot>fastboot oem unlock BYU6V9EKEGCFU5GQ
...
(bootloader)  Necessary to unlock FRP!
(bootloader)  Navigate to Developer options, and enable "OEM unlock"!
FAILED (remote: Command not allowed)
finished. total time: 0.028s

C:\Program Files (x86)\Minimal ADB and Fastboot>

**Reply by ThomasHeinrichSchmidt** [reply]
Oh, I'm sorry, that does not look promising. 
I don't think it's possible to "_use Fastboot to toggle OEM UNLOCK to "YES"_" as the (bootloader) itself says "_Navigate to Developer options, and enable "OEM unlock"!_".

**Reply by affineer** [reply]
What is the solution then?
can you please search and send the link to official Huawei Stock Firmware for p20 lite (ANE-LX2)). I have downloaded few of them, but non of them seems official one, I can not flash any firmware.

If I could get my device up and running, then I can easily unlock its bootloader, I have the bootloader unlock key now, Thanks for the PotatoNV and support.


Thanks.

**Reply by ThomasHeinrichSchmidt** [reply]
Yes, maybe you can install a recovery image, but I have never done this myself.
I would search under [easy-firmware.com](https://easy-firmware.com/index.php?a=downloads&b=search&desc=1&keyword=ANE-LX2) for the firmware.

**Reply by affineer** [reply]
flashing recovery is not allowed (loacked bootloader)

**Reply by mashed-potatoes** [reply]
> hello, it tried. but could NOT succeed. Here is the potatoNv log.
> 
> BYU6V9EKEGCFU5GQ
> 
> PotatoNV v2.0.1
> User manual: https://kutt.it/pnv-en
> --> Flashing bootloader
> Verifying images...
> Verification passed!
> Uploading hisi659a bootloader
> 
> * xloader
> * fastboot
>   Bootloader uploaded
>   Waiting for any device...
>   --> Reading information
>   Connecting to fastboot device...
> * Serial number: WCY4C18603001426
> * Board ID: 4C4XNP1862004346
> * Model: ANE-LX2
> * Build number: ANE-LX2 9.1.0.353(ZAFC185E4R1P8)
> * FBLOCK state: unlocked
>   --> Updating NVME
> * Writing FBLOCK state
>   Failed to set prop:
> 
> Where am I wrong?

Oof, it's kinda outdated but you can try the new version with some workarounds.
For some reason, the program doesn't work with this model.

**Comment by pancakeslp** [comment]
FYI, I noped out, when I saw the wires and the multimeter.  As I assumed like some of the other bootloader unlockers that you would need to buy some proprietary tech.

I would like to offer this youtube video I found that showed it doesnt need to be complicated (well more complicated than dismantling a phone)
https://www.youtube.com/watch?v=YkGugQ019ZY
(although I respect the formal advice would include multimeter and proper grounding utility)


Thanks for the app!

**Reply by ThomasHeinrichSchmidt** [reply]
I watched the video and must admit that tweezers are sufficient for grounding the testpoint. I was lucky that my testpoint was so much more accessible. The music could perhaps be a bit better.

**Reply by mcastagnaa** [reply]
That is correct. Just anything metallic on the testpoint will do. Works like a charm.

**Comment by rock3tsprocket** [comment]
When I tried using the **official** guide with my CAM-L21 (I used the Kirin 620 bootloader just in case you need confirmation I was using the wrong one) and every time I try power cycling the phone and using PotatoNV it gives me this:
 `PotatoNV v2.2.1
User manual: https://kutt.it/pnv-en
Verifying images...
Uploading hisi620...
- xloader
- fastboot
The write timed out.`
I might've done something wrong, I don't know what it is, help me please.

**Reply by ThomasHeinrichSchmidt** [reply]
Maybe try my **unofficial** guide, it may be out of date but worth a try if nothing else works.

**Reply by rock3tsprocket** [reply]
I tried it a couple of times, and now it just says 
PotatoNV v2.2.1
User manual: https://kutt.it/pnv-en
Verifying images...
The requested resource is in use.
I tried unplugging the battery and plugging it back in while the phone was disassembled but it still does that.

**Reply by ThomasHeinrichSchmidt** [reply]
Sounds like you have to start all over again.

**Reply by rock3tsprocket** [reply]
I tried an older version of the program (2.1.0, I couldn't try the same version of the program as yours due to it not supporting the Kirin 620 SoC) and it still gives the first error of the write timing out.
Edit: It now doesn't boot it seems like, but ADB still works.
Edit 2: You know how I mentioned ADB works when the device is off? I used that to boot the device.
Edit 3: Nevermind I tried the 620 CAM bootloader and it worked on the 2022.03 version.

**Comment by OverloadedTech** [comment]


**Reply by ThomasHeinrichSchmidt** [reply]
I am sorry, all I've used was the HUAWEI P8 lite 2017 = PRA-LX1. 
The procedure definitely removes all data from the phone, so do not try to unlock if you still need the data on the phone.
You could re-lock the bootloader using the above unlock code, but I didn't try at the time.

**Reply by OverloadedTech** [reply]
Actually are you sure that data will be deleted? Doesn't that depend also if FLOCK is picked or not in the tool? In my case I would need to retain them

**Reply by ThomasHeinrichSchmidt** [reply]
Are you talking about FBLOCK the flag to un/lock the bootloader? Data will be deleted as security measure against unauthorized access.

**Reply by OverloadedTech** [reply]
No, just for unlock
When you open the program there is that flag that you can toggle on and off and that should be the difference
